File: app/apis/car_imagery.py
from xml.etree import ElementTree

import requests

from .exceptions import ApiError
from ..schemas.vin import Vin

class CarImageryApiError(ApiError):
  pass

def find_car_photo_url(make: str, model: str, model_year: str) -> str | None:
  """ https://www.carimagery.com/api.pdf """
  # Make sure arguments aren't empty or whitespaces
  make, model, model_year = make.strip(), model.strip(), model_year.strip()
  if make == '' or model == '' or model_year == '':
    raise ValueError('All arguments must not be empty string and not contain only whitespaces.')

  search_term = f'{make} {model} {model_year}'
  response = requests.get(
    f'https://www.carimagery.com/api.asmx/GetImageUrl?searchTerm={search_term}',
    timeout=3
  )

  try:
    response.raise_for_status()
    root = ElementTree.fromstring(response.text)
    url = root.text
    return url if Vin.is_url(url) else None
  except ElementTree.ParseError:
    return None
  except requests.HTTPError as ex:
    raise CarImageryApiError(
      'Failed to get photo url from CarImagery API.',
      response.status_code
    ) from ex

File: nodejs/src/apis/car-imagery.ts
import { InvalidArgumentError } from '../errors/argument-error';
import xmldoc from 'xmldoc';

export class CarImageryApiError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'CarImageryApiError';
  }
}

export async function findCarPhotoUrl(make: string, model: string, modelYear: string): Promise<string | null> {
  [make, model, modelYear] = make.trim(), model.trim(), modelYear.trim();

  if (make === '' || model === '' || modelYear === '') {
    throw new InvalidArgumentError('All arguments must not be empty string and not contain only whitespaces.');
  }

  const searchTerm = `${make} ${model} ${modelYear}`;

  const url = `https://www.carimagery.com/api.asmx/GetImageUrl?searchTerm=${searchTerm}`;
  const response = await fetch(url, { 
    method: 'GET',
    headers: { "Content-Type": "application/xml",
    },
  });

  if (!response.ok) {
    throw new CarImageryApiError(`Failed to get photo url from CarImagery API. Status ${response.status} received.`);
  }

  // Expect the payload to look like this
  //   <?xml version="1.0" encoding="utf-8"?>
  //   <string xmlns="http://carimagery.com/">http://www.regcheck.org.uk/image.aspx/@dCBvIHk=</string>
  const responseText = await response.text();
  const document = new xmldoc.XmlDocument(responseText);
  return document.val;
}
